# Data Preprocessing - Bank Loan Analytics

**Purpose:** Clean and prepare the loan data for analysis.

**Approach:** Document every decision with business justification.

In [1]:
# Import Libraries
import pandas as pd
import numpy as np

# Display Settings for better readability
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

## 1. Load Data and Initial Inspection 

In [3]:
# Loading raw data
df = pd.read_csv('../data/raw/loan_data.csv')

print(f'Dataset Shape: {df.shape[0]} rows, {df.shape[1]} columns')

Dataset Shape: 45000 rows, 14 columns


In [4]:
df.head()

person_age person_gender person_education  person_income  person_emp_exp  \
0          22        female           Master          71948               0   
1          21        female      High School          12282               0   
2          25        female      High School          12438               3   
3          23        female         Bachelor          79753               0   
4          24          male           Master          66135               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT      35000    PERSONAL          16.02   
1                   OWN       1000   EDUCATION          11.14   
2              MORTGAGE       5500     MEDICAL          12.87   
3                  RENT      35000     MEDICAL          15.23   
4                  RENT      35000     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49                           3           561   
1                 0.08                           2           504   
2                 0.44                           3           635   
3                 0.44                           2           675   
4                 0.53                           4           586   

  previous_loan_defaults_on_file  loan_status  
0                             No            1  
1                            Yes            0  
2                             No            1  
3                             No            1  
4                             No            1

## 2. Data Type Assessment

**Why this matters:**
- Wrong data types can cause calculation errors.
- Categorical columns need to be identified.
- Helps us understand what kind of analysis is possible.

In [6]:
#Checking data types for each column
df.dtypes

person_age                          int64
person_gender                         str
person_education                      str
person_income                       int64
person_emp_exp                      int64
person_home_ownership                 str
loan_amnt                           int64
loan_intent                           str
loan_int_rate                     float64
loan_percent_income               float64
cb_person_cred_hist_length          int64
credit_score                        int64
previous_loan_defaults_on_file        str
loan_status                         int64
dtype: object

In [7]:
# Check for missing values
print('---Missing Values---')
print(df.isnull().sum())

---Missing Values---
person_age                        0
person_gender                     0
person_education                  0
person_income                     0
person_emp_exp                    0
person_home_ownership             0
loan_amnt                         0
loan_intent                       0
loan_int_rate                     0
loan_percent_income               0
cb_person_cred_hist_length        0
credit_score                      0
previous_loan_defaults_on_file    0
loan_status                       0
dtype: int64


In [10]:
# Checking for unique values in categorical cols
categorical_cols = ['person_gender','person_education','person_home_ownership','loan_intent','previous_loan_defaults_on_file']

print('---Unique Values in Categorical Columns---')
for col in categorical_cols:
    print(f'\n{col}:')
    print(f" Values: {df[col].unique().tolist()}")
    print(f' Count: {df[col].nunique()} unique values')

---Unique Values in Categorical Columns---

person_gender:
 Values: ['female', 'male']
 Count: 2 unique values

person_education:
 Values: ['Master', 'High School', 'Bachelor', 'Associate', 'Doctorate']
 Count: 5 unique values

person_home_ownership:
 Values: ['RENT', 'OWN', 'MORTGAGE', 'OTHER']
 Count: 4 unique values

loan_intent:
 Values: ['PERSONAL', 'EDUCATION', 'MEDICAL', 'VENTURE', 'HOMEIMPROVEMENT', 'DEBTCONSOLIDATION']
 Count: 6 unique values

previous_loan_defaults_on_file:
 Values: ['No', 'Yes']
 Count: 2 unique values


## 3. Statistical Summary - Spotting Outliers

Looking at min, max, and mean to identify suspicious values.

In [12]:
# Statistical Summary of Numerical Columns
df.describe()

person_age  person_income  person_emp_exp     loan_amnt  \
count  45000.000000   4.500000e+04    45000.000000  45000.000000   
mean      27.764178   8.031905e+04        5.410333   9583.157556   
std        6.045108   8.042250e+04        6.063532   6314.886691   
min       20.000000   8.000000e+03        0.000000    500.000000   
25%       24.000000   4.720400e+04        1.000000   5000.000000   
50%       26.000000   6.704800e+04        4.000000   8000.000000   
75%       30.000000   9.578925e+04        8.000000  12237.250000   
max      144.000000   7.200766e+06      125.000000  35000.000000   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  \
count   45000.000000         45000.000000                45000.000000   
mean       11.006606             0.139725                    5.867489   
std         2.978808             0.087212                    3.879702   
min         5.420000             0.000000                    2.000000   
25%         8.590000             0.070000                    3.000000   
50%        11.010000             0.120000                    4.000000   
75%        12.990000             0.190000                    8.000000   
max        20.000000             0.660000                   30.000000   

       credit_score   loan_status  
count  45000.000000  45000.000000  
mean     632.608756      0.222222  
std       50.435865      0.415744  
min      390.000000      0.000000  
25%      601.000000      0.000000  
50%      640.000000      0.000000  
75%      670.000000      0.000000  
max      850.000000      1.000000

## Outlier Investigation

Investigating suspicious values before deciding how to handle them.

In [14]:
# Investigate age outliers
print("=== PERSON AGE Analysis ===")
print(f"Min: {df['person_age'].min()}")
print(f"Max: {df['person_age'].max()}")
print(f"Mean: {df['person_age'].mean():.1f}")
print(f"\nHow many people are above 100 years old?")
print(f"Count: {len(df[df['person_age'] > 100])}")
print(f"\nAge distribution of people > 60:")
print(df[df['person_age'] > 60]['person_age'].value_counts().sort_index())

=== PERSON AGE Analysis ===
Min: 20
Max: 144
Mean: 27.8

How many people are above 100 years old?
Count: 7

Age distribution of people > 60:
person_age
61     12
62      7
63      3
64      7
65     10
66     11
67      1
69      5
70      7
73      3
76      1
78      1
80      1
84      1
94      1
109     1
116     1
123     2
144     3
Name: count, dtype: int64


In [15]:
# Investigating Employee Experience Outliers

print("=== EMPLOYMENT EXPERIENCE Analysis ===")
print(f"Min: {df['person_emp_exp'].min()}")
print(f"Max: {df['person_emp_exp'].max()}")
print(f"Mean: {df['person_emp_exp'].mean():.1f}")
print(f"\nHow many people have > 50 years experience?")
print(f"Count: {len(df[df['person_emp_exp'] > 50])}")

=== EMPLOYMENT EXPERIENCE Analysis ===
Min: 0
Max: 125
Mean: 5.4

How many people have > 50 years experience?
Count: 12


In [16]:
# Investigating Income Outliers

print("=== PERSON INCOME Analysis ===")
print(f"Min: ${df['person_income'].min():,}")
print(f"Max: ${df['person_income'].max():,}")
print(f"Mean: ${df['person_income'].mean():,.0f}")
print(f"Median: ${df['person_income'].median():,.0f}")
print(f"\nHow many people earn > $500,000?")
print(f"Count: {len(df[df['person_income'] > 500000])}")
print(f"\nHow many people earn > $300,000?")
print(f"Count: {len(df[df['person_income'] > 300000])}")

=== PERSON INCOME Analysis ===
Min: $8,000
Max: $7,200,766
Mean: $80,319
Median: $67,048

How many people earn > $500,000?
Count: 104

How many people earn > $300,000?
Count: 364


## 5. Preprocessing Decisions
Based on our investigation, we will apply the following filters:
| Column | Issue | Decision | Justification |
|--------|-------|----------|---------------|
| `person_age` | Max = 144 (impossible) | Cap at 70 | Reasonable max age for loan applicants |
| `person_emp_exp` | Max = 125 (impossible) | Cap at 50 | Max working career ~50 years |
| `person_income` | Max = $7.2M (extreme outlier) | Cap at $300,000 | Focus on typical borrowers, not millionaires |
**Approach:** We will FILTER OUT rows that exceed these limits, not modify the values.